config/e2guardian/e2guardian.php

```diff
@@ -0,0 +1,185 @@
+	}
+	conf_mount_rw();
+	if ($install_process == true){
+		update_output_window($uw);
+	}
+	if (isset($url) && is_url($url)){
+		if ($log_notice == true){
+			print "file download start..";
+			unlink_if_exists("/usr/local/pkg/blacklist.tgz");
+			exec("/usr/bin/fetch -o /usr/local/pkg/blacklist.tgz ".escapeshellarg($url),$output,$return);
+		} else {
+			//install process
+			if (file_exists("/usr/local/pkg/blacklist.tgz")){
+				update_output_window("Found previous blacklist database, skipping download...");
+				$return = 0;
+			} else {
+				update_output_window("Fetching blacklist");
+				download_file_with_progress_bar($url, "/usr/local/pkg/blacklist.tgz");
+				if (file_exists("/usr/local/pkg/blacklist.tgz")){
+					$return = 0;
+				}
+			}
+		}
+		if ($return == 0){
+			chdir (E2GUARDIAN_DIR . "/etc/e2guardian/lists");
+			if (is_dir ("blacklists.old")){
+				exec ('rm -rf '.E2GUARDIAN_DIR . '/etc/e2guardian/lists/blacklists.old');
+			}
+			rename("blacklists","blacklists.old");
+			exec('/usr/bin/tar -xvzf /usr/local/pkg/blacklist.tgz 2>&1',$output,$return);
+			if (preg_match("/x\W+(\w+)/",$output[1],$matches)){
+				if ($matches[1] != "blacklists"){
+					rename("./".$matches[1],"blacklists");
+				}
+				read_lists($log_notice);
+			} else {
+				file_notice("E2guardian - Could not determine Blacklist extract dir. Categories not updated","");
+			}
+		} else {
+			file_notice("E2guardian - Could not fetch blacklists from url","");
+		}
+	} else {
+		if ($install_process==true){
+			read_lists(false,$uw);
+		} elseif (!empty($url)){
+			file_notice("E2guardian - Blacklist url is invalid.","");
+		}
+	}
+}
+function read_lists($log_notice=true,$uw=""){
+	global $config,$g;
+	$group_type = array();
+	$dir = E2GUARDIAN_DIR . "/etc/e2guardian/lists";
+	//read e2guardian lists dirs
+	$groups = array("phraselists", "blacklists", "whitelists");
+	/ssigns know list files
+	$types = array('domains','urls','banned','weighted','exception','expression');
+
+	//lean previous xml config for e2guardian lists
+	foreach ($config['installedpackages'] as $key => $values){
+		if (preg_match("/e2guardian(phrase|black|white)lists/",$key)){
+			unset ($config['installedpackages'][$key]);
+		}
+	}
+	//find lists
+	foreach ($groups as $group){
+		if (is_dir("$dir/$group/")){
+			//read dir content and find lists
+			$lists = scandir("$dir/$group/");
+			foreach ($lists as $list){
+				if (!preg_match ("/^\./",$list) && is_dir("$dir/$group/$list/")){
+					$category = scandir("$dir/$group/$list/");
+					foreach ($category as $file){
+						if (!preg_match ("/^\./",$file)){
+							if (is_dir("$dir/$group/$list/$file")){
+								$subdir = $file;
+								$subcategory = scandir("$dir/$group/$list/$subdir/");
+								foreach ($subcategory as $file)
+									if (!preg_match ("/^\./",$file)){
+										//assign list to array
+										$type = split("_",$file);
+										if (preg_match("/(\w+)/",$type[0],$matches));
```

Comment:
The semi-colon on the end of this line makes a bug. The "if" will terminate there, and the next line will be executed regardless of the "if".
I have fixed in my code style review that I am doing now.